Instrument

Fristi opened the pull request:
As discussed with @joroKr21 on gitter.

It would be nice to have a typeclass `DescribeK` which includes the method name (and class name) for all the algebra methods plus the resulting computation. This can be used by tracing or metrics.

Pseudo example code:

```scala
trait OffsetStore[F[_]] {
  def get(name: String): F[Option[Long]]
  def set(name: String, eventId: Long): F[Int]
}

val dummy = new OffsetStore[RIO[Tracing, *]] {
  def get(name: String): RIO[Tracing, Option[Long]] = ZIO.succeed(Some(2l))
  def set(name: String, eventId: Long): RIO[Tracing, [Int] = ZIO.succeed(1)
}

val describeK = Derive.describe[OffsetStore]
val functorK = Derive.functorK[OffsetStore]

// transform the describeK algebra using a natural transformation which uses `RIO[Tracing, *]` and the method `String` 
// to create a OpenTracing span by using `.instrumented`
val tracing = new (Lambda[A => (RIO[Tracing, A], String)] ~> RIO[Tracing, *]) {
    def apply[A](fa: (RIO[Tracing, A], String)): RIO[Tracing, A] =
      fa._1.instrumented(fa._2)
  }

//instrumented algebra
val instrumented = functorK.mapK(describeK)(tracing)
```

The algebra plus macro is there, but the only thing I'm stuck at is how to create a type lambda in macro code? Maybe someone could jump in and help :) 

See:

```
Error:(14, 31) exception during macro expansion: 
scala.reflect.macros.TypecheckException: type mismatch;
 found   : OffsetStore[F$macro$2]
 required: OffsetStore[[A](F$macro$2[A], String)]
```

As you can see, the required type does match up with the given (which was `f`)

Also we need some tests.

kailuowang commented:
Thanks I like the idea. Not 100% sure about the TC name `DescribeK` , Maybe `Instrument`? not sure if we need a `K` there. 
In the meantime you can find some macro code working with type lambda here .
https://github.com/typelevel/cats-tagless/blob/c8b0e71fe33b29341df28b5cf6276747fe4a9af1/macros/src/main/scala/cats/tagless/MacroUtils.scala#L140

joroKr21 commented:
This is still on the level of `DeriveMacros` so you can take a look at `productK` for inspiration.
But instead of `a => Tuple2K[F, G, a]` we want to have `a => (F[a], String)`.

I'm also still not sure if it would be useful to have just the `a => String` version for anything.

joroKr21 commented on macros/src/main/scala/cats/tagless/DeriveMacros.scala:
```diff
@@ -490,6 +490,26 @@ class DeriveMacros(val c: blackbox.Context) {
+      val types = delegateAbstractTypes(Af, members, Af)
+      val Tuple2 = symbolOf[Tuple2[Any, Any]]
+      val methods = delegateMethods(Af, members, af) {
+        case method if method.occursInReturn(f) =>
```
We can only implement this if `method.returnType.typeSymbol == f` (same restriction as `productK`). Otherwise the signature won't match.

joroKr21 commented:
Thanks a lot @Fristi and welcome to `cats-tagless`!

I made a PR to your branch that implements the appropriate type for the algebra.

What we still need to decide:
  * Do we want a variant for kind `F[_]` (e.g. `Describe[F[_]]`)? I guess this is probably not that useful.
  * What should we call this? I agree with @kailuowang `DescribeK` is not a great name. But I'm not sure about `Instrument` either. Naming is hard. 
  * I think it would be nice to have a boolean flag `showArguments` of the macro that resolves `cats.Show` for each argument and then returns `methodName(...arguments.map(_.show))`. But that we can defer to another PR.

Otherwise next steps would be:
  * Work on adding an `@autoDescribe` macro annotation
  * Add some tests

Would you be interested to work on this further?

Fristi commented:
I added
- A simple test case, what tests case could you think of more?
- The className as well in the tuple. I think this is very useful, as algebras can have duplicate method names and therefore to discriminate it's handy to have the algebra name.

joroKr21 commented:
The class name is constant, so you can just use `classOf[Foo]` or directly the string in whatever you do later with the method names. I don't think we should add it, because it only complicates the type signature and generates more code.

About the name, let's use `Instrument` (without K). I couldn't come up with a better one.

About the test, it would be good to add a method with symbolic name.

The other question is if you want to work on an `@autoInstrument` macro annotation (see e.g. `@autoSemigroupalK`. So far we have macro annotations for each method in `Derive`. If not we can leave it for another PR.

Fristi commented:
I renamed to `Instrument` and added a case class `Instrumentation` which contains the value, algebraName and methodName for now. I also added the `@autoInstrument` annotation which seems to work, but when you add `@finalAlg` it crashes `failed with java.lang.NoClassDefFoundError: cats/tagless/tests/Lookup$`

joroKr21 commented on macros/src/main/scala/cats/tagless/DeriveMacros.scala:
```diff
@@ -490,6 +490,30 @@ class DeriveMacros(val c: blackbox.Context) {
+      val methods = delegateMethods(Af, members, af) {
+        case method if method.returnType.typeSymbol == f =>
+          val body = q"cats.tagless.Instrumentation(${method.body}, $algebraName, ${method.name.decodedName.toString})"
+          val returnType = appliedType(Instrumentation, F :: method.returnType.typeArgs.head :: Nil)
```
Just use `method.returnType.typeArgs`. It's shorter and no need to create a new list.

joroKr21 commented on macros/src/main/scala/cats/tagless/autoInstrument.scala:
```diff
@@ -0,0 +1,44 @@
+private[tagless] class autoInstrumentMacros(override val c: whitebox.Context) extends MacroUtils  {
+  import c.universe._
+
+  private def generateInvariantFor(algebraName: String)(algebraType: Tree, typeParams: Seq[TypeDef]) =
```
Please rename this method to `generateInstrumentFor`

joroKr21 commented on tests/src/test/scala/cats/tagless/tests/TestAlgebras.scala:
```diff
@@ -113,6 +113,11 @@ object KVStore {
 }
 
+@autoInstrument
+trait Lookup[F[_]] {
```
Let's move this to `object InstrumentTests`

joroKr21 commented on tests/src/test/scala/cats/tagless/tests/InstrumentTests.scala:
```diff
@@ -0,0 +1,56 @@
+
+
+
+
```
Nitpick: why so much whitespace?

joroKr21 commented on tests/src/test/scala/cats/tagless/tests/TestAlgebras.scala:
```diff
@@ -113,6 +113,11 @@ object KVStore {
 
+@autoInstrument
+trait Lookup[F[_]] {
+  def get(id: String): F[Option[Long]]
```
Would you mind adding a symbolic method here and then test the name?
Maybe something like this:
```scala
def ->(id: String): F[Option[String]]
```

Fristi replied:
This fails now on `Referring to non-existent method cats.tagless.tests.InstrumentTests$$anon$4.$$minus$greater(java.lang.String)cats.tagless.Instrumentation`

joroKr21 replied:
Again it works on CI. It's Intellij being funky with the tests, no problem otherwise.
In cases like this I just run `test` in `sbt`.

Fristi commented:
Fixed all your comments I think. Thanks for the review and help on getting the code to work! Really happy with the fast progress 🎉 

> It worked fine on CI, are you using Intellij?

Yes, any problems? :)

kailuowang commented on tests/src/test/scala/cats/tagless/tests/InstrumentTests.scala:
```diff
@@ -0,0 +1,61 @@
+
+class InstrumentTests extends CatsTaglessTestSuite {
+
+  test("Describe should put method and class name in result") {
```
nit pick: the term `Describe` and `desc` should better be renamed to be consistent with the type classname?

joroKr21 commented:
Oh I almost forgot. Do we want to add to the `syntax`?

kailuowang commented on core/src/main/scala/cats/tagless/Instrument.scala:
```diff
@@ -0,0 +1,30 @@
+
+import simulacrum.typeclass
+
+final case class Instrumentation[F[_], A](
```
Actually, can we name this `Instrumented` since it contains the computed value? By the way I am not married to the term instrument either, I am open to alternatives.

joroKr21 commented:
No rush from my side. One idea I was thinking about next was optionally capturing the arguments.

kailuowang commented:
Capturing arguments sounds very interesting. Can they be store in special types or just List[Any]?

joroKr21 commented:
Originally I thought about `List[List[String]]` if the parameters have `Show` instances, but `List[List[Any]]` could also work.

Fristi commented:
Fixed the comments I think :)

Would you mind adding the arguments at a later point? Would love to see `Instrument` this is a release soon if possible as it will me save a ton of code.

My two cents on the arguments. I would prefer arguments to store in a `Map[String, String]` or so where the argument it self has a `Show` instance available. The tricky part is, where do you look for those instances, and what if none around? Do we ignore and warn ? A few of me algebras use refinement types in the argument list. This requires a import which has `Contravariant` instance for these refinement types,

joroKr21 commented:
> Would you mind adding the arguments at a later point? Would love to see Instrument this is a release soon if possible as it will me save a ton of code.

That's also fine by me.

I think I also want to add #51 to the next release but that won't take very long.

kailuowang commented:
I just one last request, would you move this feature into a new package `cats.tagless.diagnosis` (open to better name)? and add a comment to the type class and macro: `This feature is experimental, API is likely to change`? 
Thanks very much!

dispalt commented:
I would like to say that I for one am very excited for this.

Fristi commented:
Fixed the comments
